intake/readers/readers.py

```diff
@@ -513,36 +513,117 @@ def _read(self, name, **kw):
 class LlamaServerReader(BaseReader):
     """Create llama.cpp server using local pretrained model file"""
 
-    output_instance = "intake.readers.datatypes:ChatService"
-    implements = {datatypes.GGUF, datatypes.SafeTensors}
-    imports = {"transformers"}
+    output_instance = "intake.readers.datatypes:LlamaCPPService"
+    implements = {datatypes.GGUF}
+    imports = {"requests"}
+
+    @classmethod
+    def _find_executable(cls):
+        import shutil
+
+        # executables were renamed in https://github.com/ggerganov/llama.cpp/pull/7809
+        path = shutil.which("llama-server")
+        if path is None:
+            # fallback on old name
+            path = shutil.which("server")
+        return path
+
+    @classmethod
+    def check_imports(cls):
+        imports = super().check_imports()
+        path = cls._find_executable()
+        return imports & (path is not None)
+
+    def _local_model_path(self, data, callback=None):
+        import os
+        from fsspec.core import split_protocol
+        from intake.catalog.default import user_data_dir
+
+        if callback is None:
+            from fsspec.callbacks import _DEFAULT_CALLBACK
+
+            cb = _DEFAULT_CALLBACK
+        else:
+            cb = callback("model")
+
+        protocol, _ = split_protocol(data.url)
+        if protocol is None:
+            # no protocol means local path
+            return data.url
+
+        storage_options = {} if data.storage_options is None else data.storage_options
+        cache_location = os.path.join(user_data_dir(), "llama.cpp")
+        options = {protocol: storage_options, "simplecache": {"cache_storage": cache_location}}
+        fs, path = fsspec.url_to_fs(f"simplecache::{data.url}", **options)
+
+        cached_fn = fs._check_file(path)
+        if cached_fn:
+            return cached_fn
+
+        with fs.fs.open(path) as f:
+            sha = fs._mapper(path)
+            cached_fn = os.path.join(fs.storage[-1], sha)
+
+            cb.set_size(fs.info(path)["size"])
+
+            cached = open(cached_fn, "wb")
+            chunk = True
+            try:
+                while chunk:
+                    chunk = f.read(fs.blocksize)
+                    cached.write(chunk)
+                    cb.relative_update(len(chunk))
+            finally:
+                cb.close()
+                cached.close()
 
-    def _read(self, data, log_file="out.log", **kwargs):
-        # TODO: list common options, like PORT
+            return cached_fn
+
+    def _read(self, data, log_file="llama-cpp.log", **kwargs):
+        startup_timeout = kwargs.pop("startup_timeout", 60)
```

Comment:
List these kwargs in the signature, so the defaults show up in documentation?

Reply:
done